File: ParkingLot/ParkingLot_Go/parkingSpot/parkingSpot.go
package parkingSpot

import (
	"parkingLot/vehicles"
	"sync"
)

func GetNewParkingSpot(id int, vehicleType vehicles.VehicleType) *ParkingSpot {
	return &ParkingSpot{
		id:          id,
		vehicleType: vehicleType,
	}
}

type ParkingSpot struct {
	id          int
	vehicle     vehicles.Vehicle
	vehicleType vehicles.VehicleType
}

func (p *ParkingSpot) GetId() int {
	return p.id
}

func (p *ParkingSpot) IsAvailable(mu *sync.Mutex) bool {
	mu.Lock()
	defer mu.Unlock()
	return p.vehicle == nil
}

func (p *ParkingSpot) GetType() vehicles.VehicleType {
	return p.vehicleType
}

func (p *ParkingSpot) ParkVehicle(v vehicles.Vehicle, mu *sync.Mutex) {
	mu.Lock()
	defer mu.Unlock()
	p.vehicle = v
}

func (p *ParkingSpot) UnParkVehicle(mu *sync.Mutex) {
	mu.Lock()
	defer mu.Unlock()
	p.vehicle = nil
}

func (p *ParkingSpot) GetVehicle() vehicles.Vehicle {
	return p.vehicle
}

File: ParkingLot/ParkingLot_Node/parkingSpot.ts
import { Vehicle, VehicleType } from "./vehicles/vehicle.ts";

export default class ParkingSpot {
  _spotNumber: number;
  _vehicle: Vehicle | null;
  _vehicleType: VehicleType;

  constructor(spotNumber: number, vehicleType: VehicleType) {
    this._spotNumber = spotNumber;
    this._vehicleType = vehicleType;
    this._vehicle = null;
  }

  get spotNumber() {
    return this._spotNumber;
  }

  get vehicle() {
    return this._vehicle;
  }

  get vehicleType() {
    return this._vehicleType;
  }

  set vehicle(v: Vehicle | null) {
    this._vehicle = v;
  }

  isAvailable() {
    return this.vehicle === null;
  }

  parkVehicle(v: Vehicle) {
    if (this.isAvailable() && this.vehicleType === v.vehicleType) {
      this.vehicle = v;
      return true;
    }
    return false;
  }

  unParkVehicle() {
    if (!this.isAvailable()) {
      this.vehicle = null;
      return true;
    }
    return false;
  }
}
